core/src/main/kotlin/behavior/interaction/ApplicationCommandInteractionBehavior.kt

```diff
@@ -34,3 +37,9 @@ internal fun ApplicationCommandInteractionBehavior(
         get() = supplier
 
 }
+
+public inline suspend fun ApplicationCommandInteractionBehavior.modal(title: String, customId: String,builder: ModalBuilder.() -> Unit): PublicInteractionResponseBehavior {
```

Comment:
1. ModalSubmit is not an ApplicationCommandInteraction nor a ComponentInteraction and as you said; it can't be invoked on the same type or PING (which we don't support *yet*) so I had to define it this way until we figure if it's worth adding a type for this as Modals evolve
2. 
2. Return value is right since you can either followup ephemerally or publically which ephemeral followups can't do.

Reply:
> 2. Return value is right since you can either followup ephemerally or publically which ephemeral followups can't do.


All `InteractionResponseBehavior`s can be followed up publicly and ephemerally (at least kords hierarchy allows this).

What I was wondering about:
- can `edit()` be used on the returned `InteractionResponseBehavior` ?
- can `delete()` be used on the returned `PublicInteractionResponseBehavior` ?

Reply:
I did some testing and these are the results:
```kotlin
suspend fun ApplicationCommandInteractionBehavior.doStuff() {

    val res: PublicInteractionResponseBehavior = this.modal(...)

    res.edit { ... } // doesn't work, defined on InteractionResponseBehavior

    res.delete() // doesn't work, defined on PublicInteractionResponseBehavior

    res.followUp { ... } // works, defined on InteractionResponseBehavior

    res.followUpEphemeral { ... } // works, defined on InteractionResponseBehavior
}
```
I suggest to add two new type in the `InteractionResponseBehavior` hierarchy (so far there are only `InteractionResponseBehavior`, `PublicInteractionResponseBehavior` and `EphemeralInteractionResponseBehavior`):
- `MessageInteractionResponseBehavior`:
this would be the new direct supertype of `[Public|Ephemeral]InteractionResponseBehavior` (they could also be renamed to `[Public|Ephemeral]MessageInteractionResponseBehavior`)
the `edit()` method would be moved down in the hierarchy from `InteractionResponseBehavior` to `MessageInteractionResponseBehavior`
- `ModalInteractionResponseBehavior`
this would be a new direct subtype of `InteractionResponseBehavior` that does not introduce new methods, followups are available for all `InteractionResponseBehavior`s
the `modal()` functions would naturally have this new type as their return type

Reply:
The Action and Data names have the following meaning for interactions:
* an ActionInteraction means that the user did perform an action like using a command, pressing a button or submitting a modal
* a DataInteraction means that discord requests some data from you like in autocomplete

By using `modal()` on an interaction, let's say a slash command, you get this response behavior we are talking about here and the user gets the modal popup. This response behavior is similar to the behavior you would get if you would instead send a message in response to the slash command. You can use it to create/get followups to the slash command interaction. The only difference is that you can't interact with the original message since there is none.

As you can see ActionInteractions differ differently from DataInteractions than ModalResponseBehaviors differ from e. g. SlashCommandResponseBehaviors. Therefore using Action / Data for the naming of response behaviors too is probably misleading.

Reply:
I first thought modals would be simmilar to DataInteractions, because they're used to transfer data, I wasn't aware of the  responses you send to a modal transaction

Reply:
> By using `modal()` on an interaction, let's say a slash command, you get this response behavior we are talking about here and the user gets the modal popup.

That's only the first part of the modal flow, after that you get an `ModalSubmitInteraction` that holds the submitted data and can be responded to similar to a slash command. The difference is that DataInteractions request data, modals give you data like a slash command gives you its options.

Reply:
yeah if they act differently they need a new type in the system

Reply:
I can think of `MutableInteractionResponse` and `ImmutableInteractionResponse`

Reply:
but then; Ephemeral can be edited but not deleted; which means that it's not a matter of mutation here

Reply:
I think `MessageInteractionResponseBehavior` and `ModalInteractionResponseBehavior` is fine - the first part of the name is about the type of the initial response which for the former is a Message and for the latter a Modal.

Reply:
I've used PopupInteractionResponse and MessageInteractionResponse as Modal shows a popup (dialog) and there might be more interactions like these

Reply:
I use PopupInteractionResponse for now for Modal
and MessageInteractionResponse for normal message-based interactions